onnx/cpp2py_export.cc

```diff
@@ -39,6 +39,57 @@ static std::string ProtoBytesToText(const py::bytes& bytes) {
   return ProtoToString(proto);
 }
 
+template <typename T, typename Ts = typename std::remove_const<T>::type>
+std::pair<std::unique_ptr<Ts[]>, std::unordered_map<std::string, T*>> ParseProtoFromBytesMap(
+    std::unordered_map<std::string, py::bytes> bytesMap) {
+  std::unique_ptr<Ts[]> values(new Ts[bytesMap.size()]);
+  std::unordered_map<std::string, T*> result;
+  size_t i = 0;
+  for (auto kv : bytesMap) {
+    ParseProtoFromPyBytes(&values[i], kv.second);
+    result[kv.first] = &values[i];
+    i++;
+  }
+  return make_pair(move(values), result);
+}
+
+std::unordered_map<std::string, py::bytes> CallNodeInferenceFunction(
+    OpSchema* schema,
+    const py::bytes& nodeBytes,
+    std::unordered_map<std::string, py::bytes> valueTypesByNameBytes,
+    std::unordered_map<std::string, py::bytes> inputDataByNameBytes,
+    std::unordered_map<std::string, py::bytes> inputSparseDataByNameBytes) {
+  NodeProto node{};
+  ParseProtoFromPyBytes(&node, nodeBytes);
+  // Early fail if node is badly defined - may throw ValidationError
+  schema->Verify(node);
+
+  // Convert arguments to C++ types, allocating memory
+  const auto& valueTypes = ParseProtoFromBytesMap<TypeProto>(valueTypesByNameBytes);
+  const auto& inputData = ParseProtoFromBytesMap<const TensorProto>(inputDataByNameBytes);
+  const auto& inputSparseData = ParseProtoFromBytesMap<const SparseTensorProto>(inputSparseDataByNameBytes);
+
+  // Construct inference context and get results - may throw InferenceError
+  shape_inference::InferenceContextImpl ctx(node, valueTypes.second, inputData.second, inputSparseData.second);
+  schema->GetTypeAndShapeInferenceFunction()(ctx);
+  // Verify the inference succeeded - may also throw ValidationError
```

Comment:
A question: wondering what happens in builds that don't support exceptions. I assume that the C++ code with abort. What does pybind return in this case? (I realize that this is a pre-existing issue, but just wondering whether we should mark/exclude tests in some special way for this case.)

Reply:
Looks like this isn't really going to be an issue, since `cmake_build` under Python requires the bindings to be built with exceptions:

https://github.com/onnx/onnx/blob/f6cb158986cf65feae64774d9a851e2649b94073/setup.py#L207-L208